narwhal/consensus/src/bullshark.rs

```diff
@@ -314,4 +329,46 @@ impl Bullshark {
         // Return its certificate and the certificate's digest.
         dag.get(&round).and_then(|x| x.get(&leader))
     }
+
+    /// Updates and calculates the reputation score for the current commit managing any internal state.
+    /// It returns the updated reputation score.
+    fn update_reputation_score(
+        &mut self,
+        state: &mut ConsensusState,
+        committed_sequence: &[Certificate],
+        sub_dag_index: u64,
+    ) -> ReputationScores {
+        // we reset the scores for every schedule change window.
+        // TODO: when schedule change is implemented we should probably change a little bit
+        // this logic here.
+        if sub_dag_index % self.change_schedule_every_committed_sub_dags == 0 {
+            state.last_consensus_reputation_score = ReputationScores::default()
+        }
+
+        // update the score for the previous leader. If no previous leader exists,
+        // then this is the first time we commit a leader, so no score update takes place
+        if let Some(previous_leader) = state.last_committed_leader {
+            for certificate in committed_sequence {
+                // TODO: we could iterate only the certificates of the round above the previous leader's round
+                if certificate
+                    .header
+                    .parents
+                    .iter()
+                    .any(|digest| *digest == previous_leader)
+                {
+                    state
+                        .last_consensus_reputation_score
+                        .add_score(certificate.origin(), 1);
```

Comment:
So we add points for whoever managed to vote for the leader. Is there a point where the fact that a node should have been a leader but did not commit gets punished. Shouldn't we add points also for the leader that it managed to get a vote?

Reply:
@LefKok good point. On this PR I've actually followed the approach we have discussed and based on what's here https://www.notion.so/mystenlabs/2023-04-Leader-Reputation-c9204dd9e91249d5a68175e7076f9e91 . Of course we can modify the algorithm to do both:
1) rewarding the leader when that exists
2) punish them when that not exists

but I would prefer having those as an addition on a follow up PR. Let's have a discussion about this